src/Microsoft.AspNetCore.Mvc.Core/Internal/FileStreamResultExecutor.cs

```diff
@@ -18,18 +22,69 @@ public FileStreamResultExecutor(ILoggerFactory loggerFactory)
 
         public Task ExecuteAsync(ActionContext context, FileStreamResult result)
         {
-            SetHeadersAndLog(context, result);
-            return WriteFileAsync(context, result);
+            var fileLength = 0L;
+            var rangeInfo = new(RangeItemHeaderValue range, long rangeLength)?();
+            if (result.FileStream.CanSeek)
+            {
+                fileLength = result.FileStream.Length;
+            }
+            if (result.LastModified.HasValue)
+            {
+                rangeInfo = SetHeadersAndLog(
+                    context,
+                    result,
+                    fileLength,
+                    result.LastModified.Value,
+                    result.EntityTag);
+            }
+            else
+            {
+                rangeInfo = SetHeadersAndLog(
+                    context,
+                    result,
+                    fileLength);
+            }
+
+            if (rangeInfo.HasValue)
+            {
+                return WriteFileAsync(context, result, rangeInfo.Value.range, rangeInfo.Value.rangeLength);
+            }
+
+            return WriteFileAsync(context, result, null, 0);
         }
 
-        private static async Task WriteFileAsync(ActionContext context, FileStreamResult result)
+        private static async Task WriteFileAsync(ActionContext context, FileStreamResult result, RangeItemHeaderValue range, long rangeLength)
         {
             var response = context.HttpContext.Response;
             var outputStream = response.Body;
 
             using (result.FileStream)
             {
-                await result.FileStream.CopyToAsync(outputStream, BufferSize);
+                if (range == null)
+                {
+                    await result.FileStream.CopyToAsync(outputStream, BufferSize);
+                }
+
+                else if (rangeLength == 0)
+                {
+                    return;
+                }
+
+                else
+                {
+                    try
+                    {
+                        result.FileStream.Seek(range.From.Value, SeekOrigin.Begin);
+                        await StreamCopyOperation.CopyToAsync(result.FileStream, outputStream, rangeLength, context.HttpContext.RequestAborted);
+                    }
+
+                    catch (OperationCanceledException)
+                    {
+                        // Don't throw this exception, it's most likely caused by the client disconnecting.
+                        // However, if it was cancelled for any other reason we need to prevent empty responses.
+                        context.HttpContext.Abort();
```

Comment:
Hmm, I'm not sure why are we adding this code, this seems like a different feature/bug